src/operations/blas2/spmv.hpp

```diff
@@ -0,0 +1,284 @@
+      loc_x[l_idx] = read_it ? vector_.eval(x_idx) : value_t(0);
+    }
+
+    const index_t I_offset = gid * x_range;
+    value_t *A = A_J_offset + I_offset;
+    index_t stride = _mat_initial_stride(J);
+
+#pragma unroll
+    for (index_t _j = 0; _j < y_chunck; ++_j) {
+      const index_t j = l_y_offset + _j;
+      const bool read_it = is_upper ? (J + _j < N) : (I_offset + l_idx < N);
+
+      if ((!is_upper && l_idx > j) || (is_upper && l_idx < j))
+        loc_A[loc_lda * l_idx + j] = loc_A[loc_lda * j + l_idx] =
+            read_it ? *A : value_t(0);
+
+      if (l_idx == j) loc_A[loc_lda * j + l_idx] = read_it ? *A : value_t(0);
+
+      A += _mat_next_stride(stride);
+    }
+  }
+
+  // ------------------------------------------------------------------------ //
+
+  // VERTICAL
+
+  {
+    value_t priv_A[y_chunck];
+
+    // this stores the matrix in local memory
+    for (index_t b = (is_upper ? 0 : gid + 1); b < (is_upper ? gid : nblock);
+         ++b) {
+      ndItem.barrier(cl::sycl::access::fence_space::local_space);
+#pragma unroll
+      for (index_t _j = 0; _j < y_chunck; ++_j) {
+        const index_t j = l_y_offset + _j;
+        priv_res += loc_A[loc_lda * j + l_idx] * loc_x[j];
+      }
+
+      const index_t I_offset = b * x_range;
+      value_t *A = A_J_offset + I_offset;
+      index_t stride = _mat_initial_stride(J);
+
+      // row full-blocks
+#pragma unroll
+      for (index_t _j = 0; _j < y_chunck; ++_j) {
+        const index_t j = l_y_offset + _j;
+        const bool read_it = is_upper ? (J + _j < N) : (I_offset + l_idx < N);
+        priv_A[_j] = read_it ? *A : value_t(0);
+        A += _mat_next_stride(stride);
+      }
+
+      ndItem.barrier(cl::sycl::access::fence_space::local_space);
+
+      if (!l_idy) {
+        const index_t x_idx = b * x_range + l_idx;
+        const bool read_it = is_upper ? true : (x_idx < N);
+        loc_x[l_idx] = read_it ? vector_.eval(x_idx) : value_t(0);
+      }
+
+#pragma unroll
+      for (index_t _j = 0; _j < y_chunck; ++_j) {
+        const index_t j = l_y_offset + _j;
+        loc_A[loc_lda * l_idx + j] = priv_A[_j];
+      }
+    }
+  }
+
+  ndItem.barrier(cl::sycl::access::fence_space::local_space);
+
+#pragma unroll
+  for (index_t _j = 0; _j < y_chunck; ++_j) {
+    const index_t j = l_y_offset + _j;
+    priv_res += loc_A[loc_lda * j + l_idx] * loc_x[j];
+  }
+
+  ndItem.barrier(cl::sycl::access::fence_space::local_space);
+
+  loc_A[loc_lda * l_idy + l_idx] = priv_res;
+
+  ndItem.barrier(cl::sycl::access::fence_space::local_space);
+
+  if (!l_idy) {
+    value_t res = value_t(0);
+#pragma unroll
+    for (index_t _j = 0; _j < y_range; ++_j) res += loc_A[loc_lda * _j + l_idx];
+
+    const index_t lhs_idx = gid * x_range + l_idx;
+    if (lhs_idx < N)
+      lhs_.eval(lhs_idx) =
+          AddOperator::eval(ProductOperator::eval(alpha_, res),
+                            ProductOperator::eval(beta_, lhs_.eval(lhs_idx)));
+  }
+
+  return 0;
```

Comment:
I think it would be better if we return the actual output here instead of `0`. If someone decides to use this node in a tree than the logic will break as we will get only `0` as output for all the work items involved.

Reply:
Done.